### DISCUSSION TITLE
zephyr prometheus library

### DISCUSSION THREAD

**Discussion by mustafaabdullahk** [zephyrproject-rtos/zephyr#72127] [Show and tell]
# Hey Zephyr Community! 🚀 

We're excited to announce the launch of our latest addition to the Zephyr ecosystem: the Zephyr Prometheus Library! 🎉

## What is Zephyr Prometheus Library?
Zephyr Prometheus is a lightweight library designed to seamlessly integrate Prometheus metrics into your Zephyr projects. Whether you're developing for IoT, edge computing, or any other embedded application with Zephyr, monitoring and collecting metrics is now easier than ever.

## Features:
📊 Effortless Integration: With just a few lines of code, you can start exposing custom metrics from your Zephyr applications.

🔍 Flexible Metrics Collection: Monitor crucial aspects of your application's performance, such as memory usage, task execution times, and custom-defined metrics.

🔒 Secure Communication: Ensure the integrity and confidentiality of your metrics data with secure communication protocols supported by Zephyr.

## How to Get Started:

Check out the Library: Dive into the [Zephyrus GitHub repository](https://github.com/mustafaabdullahk/zephyrus/) to explore the library's documentation, API reference, and examples.
Example Application: We've prepared a comprehensive example application to demonstrate the library's capabilities in action. Head over to the [Zephyr Prometheus Example repository](https://github.com/mustafaabdullahk/zephyr-prometheus-example) to get started.
Contribute and Collaborate: We welcome contributions, feedback, and ideas from the community to improve and extend the library further. Feel free to open issues, submit pull requests, or join the discussion on our GitHub repositories.
Get Involved:
Join us in revolutionizing metrics monitoring for Zephyr-powered projects! Whether you're a seasoned Zephyr developer or just getting started, the Zephyr Prometheus Library empowers you to gain deeper insights into your applications' performance.

Let's build smarter, more reliable embedded systems together with Zephyr Prometheus!

Happy coding! 🛠️

**Comment by jukkar** [comment]
Hi @mustafaabdullahk, have you considered submitting this code directly to Zephyr as the HTTP server functionality is now merged in upstream Zephyr so you could use it.

**Reply by mustafaabdullahk** [reply]
Hi @jukkar, i did follow Http Server Support pull request closely. i did start before http server support. of course i will move to zephyr http support to soon. also i want to much to integrate directly zephyr but i dont ensure how do i do it. for e.g i should write samples or develop metric library into subsys. what do you suggest for that situation?

**Comment by jukkar** [comment]
You could create a library, one option is `subsys/net/lib` if no better place can be found, that contains the "library" part of the functionality. Then a sample could use that library to show how it is used, for example the TLS certificates needed for secure connections should be placed in the sample as they clearly cannot be in the library. Then the sample could configure the library to use the given certs etc.

**Reply by mustafaabdullahk** [reply]
good idea. i will prepare merge request for that reason soon.

**Comment by mustafaabdullahk** [comment]
Hi @jukkar 

I did start implement and i close the finish. Last status of implementation library places following:

* header files places into include/zephyr/net/prometheus.
* source file places into subsys/net/lib/prometheus.
* sample app places into samples/net/prometheus.

but i've to one problem for http server. prometheus client libraries(pull method) should implement http server. that step is ok but in the zephyr http server response header start with `text/html`.

![image](https://github.com/zephyrproject-rtos/zephyr/assets/58497719/ec7d73a2-6c7b-4fa4-90ee-58466079db74)

prometheus document say following:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/58497719/35c162b9-c2af-4b2c-a91c-797455f81cfb)

how do i solve that issue?

**Comment by jukkar** [comment]
@mustafaabdullahk, Please take a look at https://github.com/zephyrproject-rtos/zephyr/pull/72629 for specifying the Content-Type field.

**Reply by mustafaabdullahk** [reply]
@jukkar that pr will solve my issue. I will merge temporarily branch and test it.

**Reply by mustafaabdullahk** [reply]
Hi @jukkar, i did create pull request: https://github.com/zephyrproject-rtos/zephyr/pull/72804

but we have one problem about http response header. when i was trying to last changes about http server, i face to response header length problem. i did save screen following issue:

![zephyr-http-server-len-issue](https://github.com/zephyrproject-rtos/zephyr/assets/58497719/05977920-b5cf-48c7-aa53-44a8e2767e56)

but header should be:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/58497719/4be0749d-efee-48a8-b2f9-7b1a5bd27b0f)


that block maybe contain strlen bug:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/58497719/db13a4c0-9f00-497d-ab8c-13c4626df910)

**Reply by jukkar** [reply]
> but we have one problem about http response header. when i was trying to last changes about http server, i face to response header length problem. i did save screen following issue:

Could you create a proper issue for this so it is not forgotten, the Discussions is not really a proper forum for bug reports.
